### PR TITLE
Add APIs for reusing

### PR DESCRIPTION
**Pull request by g41797**
Added new APIs in accordance with #19

### CODE REVIEW

File: stringbuilder.go [hunk 1]
```diff
@@ -336,3 +336,38 @@ func createTrimSet(chars ...rune) map[rune]bool {
 
 	return trimSet
 }
+
+// Set initial position
+func (s *StringBuilder) Rewind() {
```
Comment (linkdotnet) [inline stringbuilder.go]: `Clear()`does exactly that - so this method is a duplicate.

File: stringbuilder.go [hunk 2]
```diff
@@ -336,3 +336,38 @@ func createTrimSet(chars ...rune) map[rune]bool {
...
+	if index > s.position {
+		return fmt.Errorf("index cannot be greater than current position")
+	}
+	s.data[index] = val
+
+	return nil
+}
+
+// Change current position
+func (s *StringBuilder) Skip(forward int) error {
```
Comment (linkdotnet) [inline stringbuilder.go]: Can you give me an example of where this function would be useful?

Imagine we have a string "Hello" and the builder can hold 8 characters. Calling Skip(2) would work but will return a string where the last two characters are either undefined or just empty.

File: stringbuilder.go [hunk 3]
```diff
@@ -336,3 +336,38 @@ func createTrimSet(chars ...rune) map[rune]bool {
...
 	return trimSet
 }
+
+// Set initial position
+func (s *StringBuilder) Rewind() {
+	s.position = 0
+}
+
+// Sets the rune at the specific position
+func (s *StringBuilder) SetRuneAt(index int, val rune) error {
```
Comment (linkdotnet) [inline stringbuilder.go:341]: That function makes sense.

**Review by linkdotnet — CHANGES_REQUESTED** [review]
Hey @g41797,
thanks for providing a PR. 

There are some open points (see comments). There is also a **CHANGELOG.md** that describes which changes you did, which I use for new releases. So new functions should be listed there as well.

Also a few unit tests would be great. If you need help, let me know.

**Comment by g41797** [comment]
Regarding Skip - I am going to use string builder for saving parts of syslog message
Start of sb - "array of positions" - const length
The rest sets of parts
Clear+Skip allows append after header
Access to sb will be provided via wrapper - without direct access to sb methods
So I don't care

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,7 +4,12 @@ All notable changes to **ValueStringBuilder** will be documented in this file. T
 
 <!-- The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) -->
 
-## [Unreleased]
+## [Unreleased] - 2023-10-19
```
Comment (linkdotnet) [inline CHANGELOG.md]: The date will be filled by the github action:
```suggestion
## [Unreleased]
```

**Comment by linkdotnet** [comment]
> Regarding Skip - I am going to use string builder for saving parts of syslog message Start of sb - "array of positions" - const length The rest sets of parts Clear+Skip allows append after header Access to sb will be provided via wrapper - without direct access to sb methods So I don't care

For that specific scenario: Isn't `InsertAt` sufficient here?

I am a bit reluctant to add APIs, that have to be maintained, which are more of the exotic side of things and can also be modeled with the current approach. That is why I am trying to understand what is and isn't possible with the current API.

**Comment by g41797** [comment]
Insert really inserts(changes) the data
Skip doesn't change the data, only position
Scenario - reusing the same sb for parts of syslog message: [RFC3164](https://github.com/g41797/syslogsidecar#rfc3164) and [RFC5424](https://github.com/g41797/syslogsidecar#rfc5424)

**Comment by linkdotnet** [comment]
Okay, thanks for the clarification.
From an API point of view, that seems rather specialized and not something a general `StringBuilder` should handle. My biggest complaint is that you are only allowed to move forward, but you can never return.

If you need this behavior right now, you can use the `Substring` method. Is this not enough?

**Comment by g41797** [comment]
Substring returns data, but I need to append data (overwrite) from any position
I understand your doubts, but I need this functionality
I will use your forked repo with my changes
Thank you for your efforts
Do I need to close PR and #19 ?

**Comment by linkdotnet** [comment]
> Substring returns data, but I need to append data (overwrite) from any position I understand your doubts, but I need this functionality I will use your forked repo with my changes Thank you for your efforts Do I need to close PR and #19 ?

Okay - now I understand the whole picture. Sorry that I missed the bigger picture. 

That said, here two things to consider:
1. On `v1` we want to make the StringBuilder thread-safe. So you might want to "fork" from there if this feature is something you need.
2. Imagine this:
```golang
sb := NewStringBuilderFromString("FooBar")
sb.Clear() 
sb.Skip(1)
sb.Append("Baz")
```

will return a string of length 4: "FBaz". That said, that may or may not be the intended behavior for you.
But maybe a user wants `FBazar` as a result here!? That is why I am reluctant to add the API in that shape - it is very loose in its meaning.

A better approach might be something like a Cursor API:
```golang
type UnsafeCursor struct {
    builder  *StringBuilder
    position int
}

// Move moves the UnsafeCursor by the specified number of positions. Positive values move forward, 
// negative values move backward. Maybe we could use absolut positions instead. 
func (c *UnsafeCursor) Move(offset int) error {
    newPos := c.position + offset

    if newPos < 0 || newPos >= len(c.builder.data) {
        return fmt.Errorf("cursor out of bounds")
    }

    c.position = newPos
    return nil
}

// OverrideWith replaces the content of the StringBuilder from the cursor's current position 
// with the provided text.
func (c *UnsafeCursor) OverrideWith(text string) {
    textRunes := []rune(text)
    copyLen := len(textRunes)

    // Make sure there's enough space
    if c.position+copyLen > len(c.builder.data) {
        newSize := c.position + copyLen
        newData := make([]rune, newSize)
        copy(newData, c.builder.data)
        c.builder.data = newData
    }

    copy(c.builder.data[c.position:], textRunes)
    c.position += copyLen
    if c.position > c.builder.position {
        c.builder.position = c.position
    }
}

// GetUnsafeCursor returns a new UnsafeCursor instance for the StringBuilder starting at the current position.
func (s *StringBuilder) GetUnsafeCursor() *UnsafeCursor {
    return &UnsafeCursor{
        builder:  s,
        position: s.position,
    }
}
```

@varunu28 Do you have any thoughts on this? My current gut feeling still tells me it is way too specialized.

Another option I can totally live with it something like `ReplaceRange`

```golang
func (s *StringBuilder) ReplaceRange(start, end int, replacement string) error {
    if start < 0 || end >= s.position || start > end {
        return fmt.Errorf("invalid range")
    }

    replacementRunes := []rune(replacement)
    replacementLen := len(replacementRunes)
    rangeLen := end - start + 1

    newSize := s.position - rangeLen + replacementLen
    if newSize > len(s.data) {
        newData := make([]rune, newSize)
        copy(newData, s.data)
        s.data = newData
    }

    copy(s.data[start+replacementLen:], s.data[end+1:s.position])

    copy(s.data[start:], replacementRunes)
    s.position = newSize

    return nil
}
```

In your case start would be a given index and end basically the same as `sb.Len`. That would fit the theme the most **and** fixes your issue at hand.

**Comment by linkdotnet** [comment]
Edit: A way simpler implementation with the current API:
```golang
func (s *StringBuilder) ReplaceRange(start int, end int, text string) *StringBuilder {
	if start < 0 {
		start = 0
	}
	if end > s.position {
		end = s.position
	}
	if start > end {
		start = end
	}

	s.Remove(start, end-start)
	s.Insert(start, text)

	return s
}
```

That should do the trick, not @g41797 ?

**Comment by g41797** [comment]
Let me see
I need to play with you proposal

**Comment by linkdotnet** [comment]
PS: The given function would be no problem to add - one could argue we don't want to set `start` and `end` if out of bounds.

**Comment by g41797** [comment]
I am thinking about sb as storage - magnetic tape, with rune as block
With possibility to write/overwrite information  per rune or per range
Not just text, also int for length/offsets
Possibly I don't need full functionality of sb and should use part of your code as prototype
Regarding MT - possibly it's too much + serious performance hit
I take a break for thinking

**Comment by g41797** [comment]
Skip() - removed

**Comment by linkdotnet** [comment]
LGTM! Thanks!

**Review by linkdotnet — APPROVED** [review]

**Comment by linkdotnet** [comment]
Any open topics from your side, or shall I merge?

**Comment by g41797** [comment]
go ahead

**Pull request merged**